### ISSUE TITLE
False "not accessed" positive 

### ISSUE THREAD

**Issue by memeplex** [microsoft/pylance-release#1096]
## Environment data

-   Language Server version: 2021.3.3
-   OS and version: MacOS Big Sur 11.2
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.6

## Expected behaviour

In the following code fragment:

```python
        hess = np.eye(n + 1)
        res = sp.optimize.minimize(
            lambda w: 0.5 * ((w - weights) ** 2).sum(),
            weights,
            jac=lambda w: w - weights,
            hess=lambda w: hess,
            constraints=[sp.minimize.LinearConstraint(A, lb, ub)],
        )
        if not res.success:
            raise OptimizationError(res)
        weights = iter(res.x)
        c.weight = next(weights)
        for g in c.groups:
            g.weight = next(weights)
```

pylance is showing `"hess" is not accessed` for the. first line. The problem is related to the way I access the results of `minimize`: as an iterator on `res.x`. This seems to make pylance  believe I'm not using the result at all. 

## Actual behaviour

Since I'm actually using the results of `minimize` and `hess` is indeed part of the input of `minimize`, the error is spurious and should not be shown.

**Comment by memeplex** [comment]
Indeed, removing:

```python
        weights = iter(res.x)
        c.weight = next(weights)
        for g in c.groups:
            g.weight = next(weights)
```

suppresses the spurious report. If I also remove:

```python
        if not res.success:
            raise OptimizationError(res)
```

the report is now for `res`, which is all right since `res` is not used anymore.

I think the problem has to do with the assignment to `weights` in `weights = iter(res.x)`. Using `weights = res.x` or assigning to a new variable like `weights_ = iter(res.x)` both remove the warning. It's really weird.

**Comment by memeplex** [comment]
Here is a reproducible example:

```python
import numpy as np
import scipy as sp


def f():
    weights = np.ones(10)
    hess = np.eye(10)
    res = sp.optimize.minimize(
        lambda w: 0.5 * ((w - weights) ** 2).sum(),
        weights,
        jac=lambda w: w - weights,
        hess=lambda w: hess,
    )
    weights = iter(res.x)
    return weights
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.3.4, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202134-31-march-2021

**Comment by memeplex** [comment]
Thanks!
